File: 백준/Gold/1717. 집합의 표현/집합의 표현.cc
#include <stdio.h>
#include <iostream>
#include <algorithm>

using namespace std;

int p[1000001];

int find(int n) {
	if (p[n] < 0) return n;
	p[n] = find(p[n]);
	return p[n];
}

void merge(int a, int b) {
	a = find(a);
	b = find(b);
	if (a == b) return;
	p[a] += p[b];
	p[b] = a;
}

int main()
{
	int n, m, in, a, b;
	scanf("%d%d", &n, &m);
	//memset(p, -1, sizeof(int)*n);
	for (int i = 0; i <= n; i++)
		p[i] = -1;
	while (m--) {
		scanf("%d%d%d", &in, &a, &b);
		if (in) {
			if (find(a) == find(b)) printf("YES\n");
			else printf("NO\n");
		}
		else
			merge(a, b);
	}
	return 0;
}

File: 백준/Gold/1717. 집합의 표현/집합의 표현.java
import java.io.*;
import java.util.*;

public class Main {
	static int n;
	static int p[];
	
	static int find(int a) {
		if(p[a] < 0) return a;
		return p[a] = find(p[a]);
	}
	
	static void merge(int a, int b) {
		a = find(a);
		b = find(b);
		if(a == b) return;
		p[a] = b;
	}
    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringTokenizer st = new StringTokenizer(br.readLine());
        StringBuilder sb = new StringBuilder();
        n = Integer.parseInt(st.nextToken());
        p = new int[n + 1];
        for(int i = 0; i <= n; i++)
        	p[i] = -1;
        int m = Integer.parseInt(st.nextToken());
        while(m-- > 0) {
        	st = new StringTokenizer(br.readLine());
        	if(Integer.parseInt(st.nextToken()) == 0)
        		merge(Integer.parseInt(st.nextToken()), Integer.parseInt(st.nextToken()));
        	else {
        		if(find(Integer.parseInt(st.nextToken())) == find(Integer.parseInt(st.nextToken())))
        				sb.append("YES\n");
        		else
        			sb.append("NO\n");
        	}
        }
        System.out.print(sb);
    }
}